### PR TITLE
Forms material

### PR DESCRIPTION
**Pull request by xbenyx**
Seeking the integration of basic forms into a unified component while shifting to Angular Material for the administration of dynamic forms. Leveraging a metadata service and a dynamic form component, we organize each form based on type, name, label, and unique form elements, enabling the rendering of multiple forms within a single component.

Furthermore, there's a need for updated styling, aligning with the transition to Angular Material. As of now, approximately 20 forms have been adjusted to conform to Angular Material standards. The ongoing development of the component involves the continuous addition of forms to the dynamic form builder.

### CODE REVIEW

File: src/app/shared/form/form.component.ts [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+      });
+    }
+  }
+
+  /**
+   * Modifies the form values as needed for custom form submission.
+   * @param formValues - The form values to be modified.
+   * @returns The modified form values.
+   */
+  onCustomForm(formValues: any) {
```
Comment (itbf9) [inline src/app/shared/form/form.component.ts:205]: The name of this function is kind of vague. `modifyFormValues` or something is probably more descriptive?

File: src/app/shared/form/form.component.ts [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+   * Emits success alerts and navigates to the appropriate route.
+   */
+  onDeleteAction() {
+    if (this.globalMetadata['deltitle']) {
+      this.getIndex();
+    }
+    this.alert.deleteConfirmation('', this.globalMetadata['deltitle']).then((confirmed) => {
+      if (confirmed) {
+        // Deletion
+        this.gs.delete(this.apiPath, this.editedIndex).subscribe(() => {
```
Comment (itbf9) [inline src/app/shared/form/form.component.ts:232]: this subscription is never unsubscribed

File: src/app/shared/form/form.component.ts [hunk 3]
```diff
@@ -0,0 +1,244 @@
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+import { ActivatedRoute, Params, Router } from '@angular/router';
+import { Component, OnDestroy, OnInit } from '@angular/core';
+
+import { AutoTitleService } from 'src/app/core/_services/shared/autotitle.service';
+import { UnsubscribeService } from 'src/app/core/_services/unsubscribe.service';
+import { AlertService } from 'src/app/core/_services/shared/alert.service';
```
Comment (itbf9) [inline src/app/shared/form/form.component.ts:7]: Should we switch to snackbar?

File: src/app/shared/form/formuisettings.component.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+   * @param route - The ActivatedRoute for retrieving route data.
+   * @param alert - The AlertService for displaying alerts.
+   * @param gs - The GlobalService for handling global operations.
+   * @param router - The Angular Router for navigation.
+  */
+  constructor(
+    private unsubscribeService: UnsubscribeService,
+    private metadataService: MetadataService,
+    private titleService: AutoTitleService,
+    private cookieService: CookieService,
```
Comment (itbf9) [inline src/app/shared/form/formuisettings.component.ts:71]: Why store form ui settings in a cookie?

File: src/app/shared/form/formuisettings.component.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    private route: ActivatedRoute,
+    private alert: AlertService,
+    private gs: GlobalService,
+    private router: Router
+  ) {
+    // Subscribe to route data to initialize component data
+    this.route.data.subscribe((data: { kind: string, path: string, type: string }) => {
```
Comment (itbf9) [inline src/app/shared/form/formuisettings.component.ts:78]: this subscription is not unsubscribed

File: src/app/shared/form/formuisettings.component.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+  /**
+   * Initial values for form fields (optional).
+   * If provided, these values are used to initialize form controls in the dynamic form.
+   * @type {any[]}
+   */
+  formValues: any[] = [];
+
+  // Subscription for managing asynchronous data retrieval
+  private mySubscription: Subscription;
```
Comment (itbf9) [inline src/app/shared/form/formuisettings.component.ts:55]: variable name

File: src/app/shared/form/mat-autocomplete/mat-autocomplete.component.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
+import { FormControl } from '@angular/forms';
+import { Observable } from 'rxjs';
+
+@Component({
+  selector: 'app-mat-autocomplete',
+  template: `
```
Comment (itbf9) [inline src/app/shared/form/mat-autocomplete/mat-autocomplete.component.ts:7]: See https://angular.io/guide/styleguide#style-05-04

File: src/app/shared/form/mat-autocomplete/mat-autocomplete.component.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+          </mat-option>
+        </mat-autocomplete>
+      </mat-form-field>
+    </div>
+  `,
+})
+export class MatAutocompleteComponent implements OnInit {
+  @Input() options$: Observable<any[]>; // An Observable emitting the options
+  @Input() placeholder: string = 'Select or search';
+  @Output() optionSelected = new EventEmitter<any>();
```
Comment (itbf9) [inline src/app/shared/form/mat-autocomplete/mat-autocomplete.component.ts:23]: We should not emit `any`-type

File: src/app/shared/grid-containers/grid-autocol.ts [hunk 9]
```diff
@@ -1,110 +1,132 @@
...
-  left: 0;
-  width: 1px; /* Width of the vertical line */
-  background-color: gray; /* Line color, you can customize this */
-}
-`]
+    <div #content [ngStyle]="getStyles()">
+      <ng-content></ng-content>
+    </div>
+  `,
+  host: {
```
Comment (itbf9) [inline src/app/shared/grid-containers/grid-autocol.ts:10]: See https://angular.io/guide/styleguide#style-06-03

File: src/app/shared/grid-containers/grid-autocol.ts [hunk 10]
```diff
@@ -1,110 +1,132 @@
...
+    '(window:resize)': 'onWindowResize($event)'
+  },
+  styles: [
```
Comment (itbf9) [inline src/app/shared/grid-containers/grid-autocol.ts:13]: See https://angular.io/guide/styleguide#style-05-04

File: src/app/shared/grid-containers/grid-autocol.ts [hunk 11]
```diff
@@ -1,110 +1,132 @@
...
+    private renderer: Renderer2,
+    private el: ElementRef
+  ) {}
 
-  ngOnInit() : void {
+  ngOnInit(): void {
     this.isMobile = this.width < this.mobileWidth;
   }
 
+  // Method to handle window resize events
```
Comment (itbf9) [inline src/app/shared/grid-containers/grid-autocol.ts:49]: We previously used JSDoc styled comments. We should use the same everywhere i think

File: src/app/shared/grid-containers/grid-autocol.ts [hunk 12]
```diff
@@ -1,110 +1,132 @@
...
-    return Math.floor((this.width + this.gutterSize) / (this.cardWidth + this.gutterSize));
+    return Math.floor(
+      (this.width + this.gutterSize) / (this.cardWidth + this.gutterSize)
+    );
   }
 
-  getRows(){
-    return  Math.floor((this.height + this.gutterSize) / (this.cardHeight + this.gutterSize));
+  // Calculate the number of rows based on available space
+  getRows() {
```
Comment (itbf9) [inline src/app/shared/grid-containers/grid-autocol.ts:64]: Its a good practice to always add return type

File: src/app/shared/navigation/horizontalnav.component.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
+import { HorizontalNav } from 'src/app/core/_models/horizontalnav.model';
+import { Component, Input, OnDestroy } from '@angular/core';
+import { takeUntil } from 'rxjs/operators';
+import { Router } from '@angular/router';
+import { Subject } from 'rxjs';
+
+@Component({
+  selector: 'horizontalnav',
+  template: `
+    <mat-toolbar>
```
Comment (itbf9) [inline src/app/shared/navigation/horizontalnav.component.ts:10]: See https://angular.io/guide/styleguide#style-05-04

File: src/app/users/users-routing.module.ts [hunk 14]
```diff
@@ -4,90 +4,98 @@ import { IsAuth } from "../core/_guards/auth.guard";
...
+              path: SERV.ACCESS_GROUPS,
               breadcrumb: 'New Access Group',
               permission: 'GroupAccess'
           },
-          canActivate: [IsAuth,CheckPerm]},
+          canActivate: [CheckPerm]},
         {
-          path: 'access-groups/:id/edit', component: CUGroupComponent,
+          path: 'access-groups/:id/edit', component: FormComponent,
           data: {
```
Comment (itbf9) [inline src/app/users/users-routing.module.ts:91]: We should probably create an interface for router data

File: src/styles/base/_base.scss [hunk 15]
```diff
@@ -111,4 +111,18 @@ h2 {
     border-style: solid;
     border-color: var(--mat-table-row-item-outline-color, rgba(0, 0, 0, 0.12));
   }
-}
\ No newline at end of file
+}
+
+// Angular Material
+
+.btn-toogle-select {
```
Comment (itbf9) [inline src/styles/base/_base.scss:118]: Maybe this should be in form css?

**Review by itbf9 — CHANGES_REQUESTED** [review]
Nice work! I had a few comments but nothing big really. Please see my comments

File: src/app/auth/auth.component.ts [hunk 16]
```diff
@@ -1,5 +1,13 @@
-import { AuthService, AuthResponseData } from '../core/_services/access/auth.service';
-import { faLock, faUser, faEye, faEyeSlash } from '@fortawesome/free-solid-svg-icons';
+import {
+  AuthResponseData,
+  AuthService
+} from '../core/_services/access/auth.service';
+import {
+  faEye,
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/auth/auth.component.ts:6]: Yes, in fact im using mat-icon. I forgot to remove the library. thanks

File: src/app/auth/auth.component.ts [hunk 17]
```diff
@@ -8,20 +16,102 @@ import { Observable } from 'rxjs';
 import { ConfigService } from '../core/_services/shared/config.service';
 
 @Component({
-  selector: 'app-auth',
-  templateUrl: './auth.component.html'
-})
-export class AuthComponent implements OnInit {
+  selector: 'app-login',
+  template: `
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/auth/auth.component.ts:20]: Yes, makes sense. Move out as before.

File: src/app/auth/auth.component.ts [hunk 18]
```diff
@@ -8,20 +16,102 @@ import { Observable } from 'rxjs';
...
+      }
+    `
+  ]
+})
+export class AuthComponent implements OnInit {
   errorRes: string | null;
 
-  public showPassword: boolean;
-  public showPasswordOnPress: boolean;
+  public hide = true;
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/auth/auth.component.ts:114]: Cool. Added some additional documentation in the Style Guide, I changed the name to isVisible as has other options for visibility_off.

File: src/app/shared/form/formuisettings.component.ts (hunk 4 above)
Comment (xbenyx) [reply src/app/shared/form/formuisettings.component.ts:71]: I did this compoment before I saw that you created one. I need to adapt it, for the moment i'll delete it.

File: src/app/shared/form/dynamicformlayout.component.ts [hunk 19]
```diff
@@ -0,0 +1,369 @@
...
+import { MetadataService } from 'src/app/core/_services/metadata.service';
+import { GlobalService } from 'src/app/core/_services/main.service';
+import { ChangeDetectorRef } from '@angular/core';
+
+/**
+ * This component renders a dynamic form based on the provided form metadata.
+ */
+@Component({
+  selector: 'app-dynamic-form',
+  template: `
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/dynamicformlayout.component.ts:15]: Moved html out

File: src/app/shared/form/dynamicformlayout.component.ts [hunk 20]
```diff
@@ -0,0 +1,369 @@
...
+
+  /**
+   * An array of form field metadata that describes the form structure.
+   */
+  @Input() formMetadata: any[] = [];
+
+  /**
+   * Additional CSS class for labels.
+   */
+  @Input() labelclass?: any;
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/dynamicformlayout.component.ts:117]: Removed this, I used this in the old UI but currently not used.

File: src/app/shared/form/dynamicformlayout.component.ts [hunk 21]
```diff
@@ -0,0 +1,369 @@
...
+  }
+  // Create the Angular FormGroup with the configured controls.
+  this.form = this.fb.group(formRows);
+  }
+
+  /**
+   * A subscription to handle dynamic select options data retrieval.
+   * This subscription is used to fetch and update select field options with dynamic data.
+   */
+  private selectOptionsSubscription: Subscription;
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/dynamicformlayout.component.ts:247]: Moved them up

File: src/app/shared/navigation/horizontalnav.component.ts (hunk 13 above)
Comment (xbenyx) [reply src/app/shared/navigation/horizontalnav.component.ts:10]: PErfect.

File: src/app/auth/auth.component.ts [hunk 22]
```diff
@@ -8,20 +16,102 @@ import { Observable } from 'rxjs';
...
+      </mat-card>
+      <mat-card class="mat-card-alert">
+        <mat-card-content>
+          This is a private closed system. If you need access, you need to
+          contact an admin.
+        </mat-card-content>
+      </mat-card>
+    </div>
+  `,
+  styles: [
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/auth/auth.component.ts:70]: This component still on progress but moved the styles to form.scss and added the sass variables

File: src/app/shared/form/dynamicformlayout.component.ts [hunk 23]
```diff
@@ -0,0 +1,369 @@
...
+
+  /**
+   * The Angular FormGroup that represents the dynamic form.
+   */
+  @Input() form: FormGroup;
+
+  /**
+   * Define the custom layout array.
+   */
+  @Input() customLayout: any[];
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/dynamicformlayout.component.ts:132]: Removed this component for the moment,

File: src/app/shared/form/dynamicformlayout.component.ts [hunk 24]
```diff
@@ -0,0 +1,369 @@
...
+   */
+  constructor(private fb: FormBuilder, private gs: GlobalService, private cd: ChangeDetectorRef) {}
+
+  /**
+   * Initializes the dynamic form by creating form controls and setting their initial values.
+   * This method is called when the dynamic form component is initialized.
+   */
+  ngOnInit() {
+  // Check if this.customLayout is iterable and fallback to an empty array if it's not.
+  const customLayout = Array.isArray(this.customLayout) ? this.customLayout : [];
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/dynamicformlayout.component.ts:184]: Removed this component for the moment.

File: src/app/shared/form/dynamicformlayout.component.ts [hunk 25]
```diff
@@ -0,0 +1,369 @@
...
+  isLoadingSelect: boolean = true;
+
+  /**
+   * Angular lifecycle hook: ngAfterViewInit
+   * Performs initialization and logic for select fields with dynamic options.
+   */
+  ngAfterViewInit() {
+    // Check if there are any "select" type fields with "selectOptions$"
+    const selectFields = this.formMetadata.filter(
+      (field) => field.type === 'selectd' && field.selectOptions$
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/dynamicformlayout.component.ts:262]: Removed this component for the moment. Selectd should be custom component.

File: src/app/shared/form/dynamicformlayout.component.ts [hunk 26]
```diff
@@ -0,0 +1,369 @@
...
+
+  /**
+   * Transforms API response options based on a field mapping configuration.
+   *
+   * @param apiOptions - The options received from an API response.
+   * @param field - The field configuration that contains the mapping between form fields and API fields.
+   *
+   * @returns An array of transformed select options to be used in the form.
+   */
+  transformSelectOptions(apiOptions: any[], field: any): any[] {
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/dynamicformlayout.component.ts:310]: Removed this component for the moment.

File: src/app/shared/form/form.component.ts [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+import { MetadataService } from 'src/app/core/_services/metadata.service';
+import { GlobalService } from 'src/app/core/_services/main.service';
+import { Subscription } from 'rxjs';
+
+@Component({
+  selector: 'app-form',
+  template: `
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/form.component.ts:14]: Done

File: src/app/shared/form/form.component.ts [hunk 28]
```diff
@@ -0,0 +1,244 @@
...
+
+  /**
+   * Initial values for form fields (optional).
+   * If provided, these values are used to initialize form controls in the dynamic form.
+   * @type {any[]}
+   */
+  formValues: any[] = [];
+
+  // Subscription for managing asynchronous data retrieval
+  private mySubscription: Subscription;
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/form.component.ts:89]: Changed

File: src/app/shared/form/form.component.ts [hunk 29]
```diff
@@ -0,0 +1,244 @@
...
+    });
+    // Add this.mySubscription to UnsubscribeService
+    this.unsubscribeService.add(this.mySubscription);
+  }
+
+  /**
+   * Loads data for editing a form.
+  */
+  getIndex() {
+    this.route.params.subscribe((params: Params) => {
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/form.component.ts:138]: Done

File: src/app/shared/form/form.component.ts [hunk 30]
```diff
@@ -0,0 +1,244 @@
...
+  /**
+   * Loads data for editing a form.
+   */
+  loadEdit() {
+    this.route.params.subscribe((params: Params) => {
+      this.editedIndex = +params['id'];
+    });
+
+    // Fetch data from the API for editing
+    this.gs.get(this.apiPath, this.editedIndex).subscribe((result) => {
```
[the remark this answers is not in the document]
Comment (xbenyx) [reply src/app/shared/form/form.component.ts:152]: Done

File: src/app/shared/form/form.component.ts (hunk 1 above)
Comment (xbenyx) [reply src/app/shared/form/form.component.ts:205]: modifyFormValues makes more sense

File: src/app/shared/grid-containers/grid-autocol.ts (hunk 11 above)
Comment (xbenyx) [reply src/app/shared/grid-containers/grid-autocol.ts:49]: That is an old component. But agree.

File: src/app/shared/grid-containers/grid-autocol.ts (hunk 12 above)
Comment (xbenyx) [reply src/app/shared/grid-containers/grid-autocol.ts:64]: Perfect, added.

File: src/styles/base/_base.scss (hunk 15 above)
Comment (xbenyx) [reply src/styles/base/_base.scss:118]: Moved to form and added their sass variables

File: src/app/shared/form/form.component.ts (hunk 3 above)
Comment (xbenyx) [reply src/app/shared/form/form.component.ts:7]: Change service?

File: src/app/users/users-routing.module.ts (hunk 14 above)
Comment (xbenyx) [reply src/app/users/users-routing.module.ts:91]: Done

File: src/app/shared/form/form.component.ts (hunk 3 above)
Comment (itbf9) [reply src/app/shared/form/form.component.ts:7]: Ok I see, you wrap snackBar in a custom service? Yes that's probably a good idea. Never mind then :)

**Review by itbf9 — APPROVED** [review]
LGTM

**Review by itbf9 — APPROVED** [review]

**Pull request merged**